conda_smithy/templates/azure-pipelines-osx.yml.tmpl

```diff
@@ -28,15 +28,45 @@ jobs:
 
 {%- if azure.store_build_artifacts %}
   - script: |
-        artifact_name="conda_artifacts_$(build.BuildId)_$(SHORT_CONFIG_NAME)"
-        echo "##vso[task.setVariable variable=ARTIFACT_NAME]$artifact_name"
-        if [ -d /Users/runner/miniforge3/conda-bld/ ]; then
-          echo "##vso[task.setVariable variable=CONDA_BLD_DIR_EXISTS]true"
+        # Set the artifact name, specialized for this particular job run.
+        ARTIFACT_NAME="conda_artifacts_$(build.BuildNumber).$(system.JobAttempt)_$(CONFIG)"
+        if [[ {% raw %}${#ARTIFACT_NAME}{% endraw %} -gt 80 ]]; then
+          ARTIFACT_NAME="conda_artifacts_$(build.BuildNumber).$(system.JobAttempt)_$(SHORT_CONFIG)"
         fi
-    displayName: Check for conda build artifacts
+        echo "##vso[task.setVariable variable=ARTIFACT_NAME]$ARTIFACT_NAME"
+        echo "ARTIFACT_NAME: $ARTIFACT_NAME"
+
+        # Check that the conda-build directory exists for archiving.
+        ARTIFACT_DIR=/Users/runner/miniforge3/conda-bld/
+        if [ -d "$ARTIFACT_DIR" ]; then
+          # remove caches to save some artifact space
+          rm -rf $ARTIFACT_DIR/git_cache
+          rm -rf $ARTIFACT_DIR/pip_cache
+          rm -rf $ARTIFACT_DIR/src_cache
+          # delete broken symlinks so that artifact publishing doesn't fail
+          find $ARTIFACT_DIR -type l -exec test ! -e {} \; -delete
+        else
+          echo "##vso[task.logissue type=error]conda-build directory does not exist"
+          exit 1
+        fi
+        echo "##vso[task.setVariable variable=ARTIFACT_PATH]$ARTIFACT_DIR"
+        echo "ARTIFACT_DIR: $ARTIFACT_DIR"
+
+        if [ $AGENT_JOBSTATUS == "Failed" ]; then
+          # Create a zip archive to use as the artifact.
+          ARTIFACT_ZIP="$(Build.ArtifactStagingDirectory)/${ARTIFACT_NAME}.zip"
+          cd "$ARTIFACT_DIR"
+          zip -r -y -q "$ARTIFACT_ZIP" . -x '*.git/*' '*_*_env*/*' '*_cache/*'
```

Comment:
Why exclude `_env*`? They are useful for debugging.

Reply:
Because they can be huge and I didn't personally see a use for debugging with it generally being possible to get that information in other places, but if you'd rather they stay then that's fine too.

Reply:
It would also be possible to create a second zip and artifact that just contains the envs, so that they are there if necessary but don't always have to be downloaded if not necessary. At least the pre-zipping makes it more manageable than currently, where I've had to wait 10s of minutes for the download to happen when the environments are big and there are lots of files that need to be pulled into their auto-generated zip.

Reply:
I updated the PR so that the conda environment directories are now stored as a second artifact when the build fails. You can see the result [here](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=335164&view=artifacts&pathAsName=false&type=publishedArtifacts). I think that should make everyone happy.